src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportJobTask.cs

```diff
@@ -227,25 +228,109 @@ private async Task UpdateJobRecordAsync(CancellationToken cancellationToken)
+                foreach (var filter in _exportJobRecord.Filters)
+                {
+                    filteredResources.Add(filter.ResourceType.ToUpper(CultureInfo.InvariantCulture));
+                }
             }
-            else if (_exportJobRecord.ExportType == ExportJobType.All && !string.IsNullOrEmpty(_exportJobRecord.ResourceType))
+
+            IAnonymizer anonymizer = IsAnonymizedExportJob() ? await CreateAnonymizerAsync(cancellationToken) : null;
+
+            if (progress.CurrentFilter != null)
             {
-                queryParametersList.Add(Tuple.Create(KnownQueryParameterNames.Type, _exportJobRecord.ResourceType));
+                await ProcessFilter(exportJobConfiguration, progress, queryParametersList, sharedQueryParametersList, anonymizer, "filter", cancellationToken);
             }
 
-            IAnonymizer anonymizer = IsAnonymizedExportJob() ? await CreateAnonymizerAsync(cancellationToken) : null;
+            if (_exportJobRecord.Filters != null && _exportJobRecord.Filters.Any(filter => !progress.CompletedFilters.Contains(filter)))
+            {
+                foreach (var filter in _exportJobRecord.Filters)
+                {
+                    if (!progress.CompletedFilters.Contains(filter) &&
+                        requestedResourceTypes != null &&
+                        requestedResourceTypes.Contains(filter.ResourceType, StringComparison.OrdinalIgnoreCase) &&
+                        (_exportJobRecord.ExportType == ExportJobType.All || filter.ResourceType.Equals(KnownResourceTypes.Patient, StringComparison.Ordinal)))
+                    {
+                        progress.SetFilter(filter);
+                        await ProcessFilter(exportJobConfiguration, progress, queryParametersList, sharedQueryParametersList, anonymizer, "filter", cancellationToken);
+                    }
+                }
+            }
+
+            // The unfiltered search should be run if there were no filters specified, there were types requested that didn't have filters for them, or if a Patient/Group level export didn't have filters for Patients.
+            // Examples:
+            // If a patient/group export job with type and type filters is run, but patients aren't in the types requested, the search should be run here but no patients printed to the output
+            // If a patient/group export job with type and type filters is run, and patients are in the types requested and filtered, the search should not be run as patients were searched above
+            // If an export job with type and type filters is run, the search should not be run if all the types were searched above.
+            if (_exportJobRecord.Filters == null ||
+                _exportJobRecord.Filters.Count == 0 ||
+                (_exportJobRecord.ExportType == ExportJobType.All &&
+                !requestedResourceTypes.All(resourceType => filteredResources.Contains(resourceType.ToUpper(CultureInfo.InvariantCulture)))) ||
```

Comment:
>!requestedResourceTypes.All [](start = 16, length = 27)

Should we check whether `requestedResourceTypes` can be null here?

Reply:
Added exception check